### PR TITLE
Fix concurrent_unordered_map to prevent accessing padding bits in pair_type

### PR DESCRIPTION
**Pull request by davidwendt**
The `concurrent_unordered_map` class uses a `thrust::pair` type internally to manage key/value pairs. There is also an also an `pair_packer` utility that will pack the pair into a single type (uint32 or uint64) to better utilize the `atomicCAS` for storing the pair. The code also includes a path for cases when the `thrust::pair` is not packable.

The `is_packable` tests to make sure both pairs are integer types and can be packed in either a single uint32 or uint64 type. Unfortunately, if the types are different sizes (e.g. `thrust::pair<uint64_t, uint32_t>`) the pair-type also contains padding bits. This resulted in a crash in `semi_left_join` which resulted in a pair type of `thrust::pair<uint32_t,bool>`. Referencing the single packed value therefore included padding bits which is undefined behavior.

<details>
  <summary>Debug session showing packed values with non-zero padding bits</summary>

[New Thread 0x7f05725fd700 (LWP 6311)]
[New Thread 0x7f0571c77700 (LWP 6312)]
[Switching focus to CUDA kernel 0, grid 5, block (0,0,0), thread (0,0,0), device 0, sm 0, warp 0, lane 0]

Thread 1 "map_test" hit Breakpoint 1, concurrent_unordered_map<int, bool, cudf::row_hasher<default_hash, true>, cudf::row_equality_comparator<true>, default_allocator<thrust::pair<int, bool> > >::attempt_insert<thrust::pair<int, bool> > (this=0x7f0574fffb08, insert_location=0x7f0542e00448, insert_pair=0x7f0574fff9f8)
    at /cudf/cpp/build/../src/hash/concurrent_unordered_map.cuh:283
283         if (old.packed == unused.packed) { return insert_result::SUCCESS; }
///// expected these 2 values to be equal but the random padding bits makes this false
(cuda-gdb) p old.packed
$1 = 6442450943
(cuda-gdb) p unused.packed
$2 = 8430733011321880575

(cuda-gdb) p old.pair.first
$3 = 2147483647
(cuda-gdb) p old.pair.second
$4 = true

(cuda-gdb) p unused.pair.first
$5 = 2147483647
(cuda-gdb) p unused.pair.second
$6 = true

(cuda-gdb) p/x unused.packed
$9 = 0x74fffb017fffffff
(cuda-gdb) p/x old.packed
$10 = 0x17fffffff

(cuda-gdb) p unused
$11 = {packed = 8430733011321880575, pair = {first = 2147483647, second = true}}
(cuda-gdb) p/x unused.pair.first
$12 = 0x7fffffff
(cuda-gdb) p/x unused.pair.second
$13 = 0x1
(cuda-gdb)

(cuda-gdb) p old
$14 = {packed = 6442450943, pair = {first = 2147483647, second = true}}
(cuda-gdb) p/x old.pair.first
$15 = 0x7fffffff
(cuda-gdb) p/x old.pair.second
$16 = 0x1
(cuda-gdb)

</details>

This PR solves this by adding a size check to the current `is_packable()`. The C++17 [std::has_unique_object_representations](https://en.cppreference.com/w/cpp/types/has_unique_object_representations) is used here to ensure no padding bits are in the `pair_type`.

### CODE REVIEW

**Comment by vuule** [comment]
rerun tests

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 1]
```diff
@@ -263,7 +265,7 @@ class concurrent_unordered_map {
    */
   template <typename pair_type = value_type>
   __device__ std::enable_if_t<is_packable<pair_type>(), insert_result> attempt_insert(
-    value_type* insert_location, value_type const& insert_pair)
+    value_type* const __restrict__ insert_location, value_type const& insert_pair)
```
Comment (vuule) [inline cpp/src/hash/concurrent_unordered_map.cuh:268]: is const relevant here?

**Review by vuule — APPROVED** [review]

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 1 above)
Comment (davidwendt) [reply cpp/src/hash/concurrent_unordered_map.cuh:268]: Not really. I just wanted to two `attempt_insert` functions defined here to match. And removing `const` from the 2nd one just seemed wrong.

Comment (vuule) [reply cpp/src/hash/concurrent_unordered_map.cuh:268]: No harm in keeping it, I was just curious if there's maybe an interaction with `__restrict__`.

**Review by jrhemstad — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**